### PR TITLE
[BEAM-22089] Fix query retry in Java FirestoreIO

### PR DESCRIPTION
**Pull request by danthev**
This fixes a condition where, when a RunQuery request is retried, the cursor inferred from the last received value is set incorrectly, potentially yielding no or skipping some results. This can happen due to nested field paths in the orderBy or when the implicit order (e.g. by __name__) is not filled correctly. See also #22089.
This PR fixes both root-causes along with unit test cases for each.

------------------------

[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

### CODE REVIEW

**Comment by danthev** [comment]
R: @yixiaoshen 
R: @clement

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java
Comment (yixiaoshen) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java:329]: nit: since the field names are using "foo", "foobar", "bar", I'd suggest changing field values to something else instead of "bar" and "foo" to avoid confusion between field names and field values.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java
Comment (yixiaoshen) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java:355]: maybe just declare two Filter constants FOO_EQUALS_BAR and FOO_NOT_EQUALS_FOO around line 297? there's no argument to these two methods so everytime these methods get called they just return the same filter, constants will do just as good as methods.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java
Comment (yixiaoshen) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java:138]: maybe also add a test for backticked fields, e.g. the query has an inequality on field "\`x.y\`" or has an order by on field "\`x.y\`", and also add a field "x.y" to the returned document in newResponse() to support this test.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java]: Sure why not. Done.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java
Comment (yixiaoshen) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java:138]: oh I didn't think about this case (that a back ticked field name is inside a nested map), I don't know what the code is going to do but seems that it's handling it just fine?

Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java:138]: Implemented escaping logic to handle these cases.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (yixiaoshen) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:165]: I feel this should ideally be added as some class in googleapis/java-firestore and imported by beam as a library, in that way we will have better control over this logic and less likely to have it diverge from our backend (not that this logic is changed that frequently though). I guess it's OK as a temporary quick fix.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (yixiaoshen) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:136]: RuntimeExceptions do not need to be declared in the throws clause.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (yixiaoshen) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java:132]: nit: can we combine these two lines?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:157]: Because the regex matched, fieldPath is either empty here, or it is `.nestedfieldname`, so I'm removing the dot. `if(!fieldPath.isEmpty())` would have the same effect, but I think it's probably less clear?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java:132]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:136]: Removed them.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:128]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:127]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (yixiaoshen) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:160]: SG, maybe add a inline comment that when fieldPath is not empty it must start with "." here.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java
Comment (yixiaoshen) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java:352]: You added a super complicated case "\`fo\\\\\`o.m\\\\\`ap\`.\`bar.key\`" and the most simple case "foo" for the order by, can we also add two test cases that are in between these two: (1) x.y as y inside the map x, both x and y are simple strings and not escaped; (2) \`x.y\` as a single property with a dot in the property name. So the document might look like this:
```
{
  "foo": "bar"
  "fo`o.m`ap": { "bar.key": "bar.val" }
  "baz" : {"qux" : "val"}
  "quux.quuz" : 123
}
```
you might parameterize the tests if it makes things easier, or just writing these 4 separate tests also seem fine.

**Comment by yixiaoshen** [comment]
LGTM. please also get a review from someone else.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1FnRunQueryTest.java]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:161]: Done.

**Comment by danthev** [comment]
R: @johnjcasey 
(yixiaoshen@ has approved, review by clem@ still pending)

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (johnjcasey) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:169]: Agreed. A lot of this looks specific to firestore internals, so it would be nice to push it down, away from user space

**Comment by johnjcasey** [comment]
Nothing looks bad to me, but a lot of this is firestore specific, so its hard to judge from a BeamIO perspective

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (johnjcasey) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java:124]: Is this general firestore knowledge? if not, can we add some sort of documentation link here for the future?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:169]: Yes, this is a good idea not just for Beam but also to make it harder for users tp shoot themselves in the foot with the Rest API. We'll have an AI to implement this there, but leave it in Beam for now.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (clement) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:127]: The fact that this (non-private) method mutates its argument is a really sharp corner. A couple of possible ways out of that:

- make it and `resolveOrderByFieldPath` private, and expose only a single method that returns the `Value` for a `Document` by field-path, with the path given as a string.
- pass the current index of the segment as an extra parameter

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (clement) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:43]: This class could use unit-tests.

**Review by clement — CHANGES_REQUESTED** [review]

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java:114]: Hmm, that reporting of partial progress sounds like a potential failure case. Could you have another look at the implementation there around line 415? I think we might have to completely override `processElement` for RunQueryFn and only set lastReceivedValue if the response contained a document.
Beam only checks if the response is null before calling `setStartFrom`, though it also only calls it if the ServerStream has ended in a RuntimeException. So I guess with partial progress response -> query times out -> no document to restart from, we'd have a problem?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java:122]: You're right, this is basically a "this should never happen". I'll remove the rethrow altogether.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:75]: Ah, nice catch. Guess I'll have to unescape all these too.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:120]: @yixiaoshen was saying that the filters in a composite filter would already be sorted. But since StructuredQuery is a raw proto, I'm guessing this would only be the case on the backend?
If not I'll add sorting to the composite filter case, and the default case should probably throw a NotImplemented instead of a break.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java]: Updated comment to be explicit that this is in the top-level valueMap.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:145]: Done, made it so RunQueryFn only calls `lookupDocumentValue(Document document, String fieldPath)`

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:44]: Will add on Monday.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (clement) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:172]: s/escaped/unescaped ?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (clement) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:128]: > that the filters in a composite filter would already be sorted.

The filter order in a composite filter is arbitrary, and picked by the user.

> If not I'll add sorting to the composite filter case

Composite filters can be nested, the sorting needs to be done for all field paths that appear at any point in the query filter "tree". I think it's also reasonable to not sort and only ensure that the results contains only one *canonical* field paths, otherwise fail with a clear message that **this** version of the beam connector does not support query with multiple inequalities (this is punting the issue for later, and hoping we will have a better mechanism for resumption by then).

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (clement) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java]: > So I guess with partial progress response -> query times out -> no document to restart from, we'd have a problem?

Yes, looks like it. Perhaps add an abstract reducer-like method to `StreamingFirestoreV1ReadFn`? Something with a signature like `OutT resumptionValue(OutT previousValue, OutT newValue)`?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java]: I went ahead and implemented a class modeled after [FieldPath](https://github.com/googleapis/java-firestore/blob/main/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FieldPath.java) with the extra escaping logic but without support for non-Where field paths. Just adding a compareTo there was easy, so the sorting is fully implemented now.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java]: Yes, done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java]: That's a great idea, just added that. 
I saw BatchGetDocuments may also return no results, `setStartFrom` will request all documents in that case anyway, but I added an appropriate resumptionValue there so we don't even iterate.

**Comment by danthev** [comment]
Code is ready again, fixed some spotless errors after adding all the unit tests, filter sorting etc.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java]: The [documentation](https://firebase.google.com/docs/firestore/reference/rest/v1/projects.databases.documents#Document) for `Document` says that __.*__ is reserved, so it sort of implies that __name__ can't be in there. I can link that here.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (clement) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:202]: The segments must be escaped for this to be valid.

Rather than also writing the code for escaping, one suggestion would be to save the escaped field path this instance was constructed from, and return it. It will not be the canonical escaped representation, but it will be valid.

A unit test should show the issue. For ex a query with no order by and an inequality filter on `foo.bar`.bar would add foo.bar.baz to the order by, which is different.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (clement) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:231]: UTF16 does not follow unicode code point ordering. Can we use something like the ICU4J library which has better support for unicode string comparison here? Or compare the UTF8 bytes.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtilsTest.java
Comment (clement) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtilsTest.java:125]: This could use some extra test case, for example:
- multiple equivalent field paths on inequalities (\`foo\` > 1 AND foo < 2)
- inequalities on `__name__` + other field
- orderby partially matching some inequalities / `__name__`
- order by matching but different escaping: WHERE \`foo\` > 1 ORDER BY foo

**Review by clement — CHANGES_REQUESTED** [review]

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java]: So I guess the Firestore SDK FieldPath implementation is wrong? I've switched it to comparing UTF8 bytes.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java]: Ah yeah, added a test case that shows it. I've switched it to returning the original string.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtilsTest.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtilsTest.java:139]: \#3 was already in there, I've added another test case for \#2 and tweaked the main one to cover \#1 and \#4.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (clement) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java:242]: This looks like a signed byte comparator, which would be wrong (try a test with an ascii character vs a UTF8 multi-byte code point).

If you have access to Guava's `UnsignedBytes.lexicographicalComparator` this can be used directly on the byte arrays returned by `getBytes`.

**Review by clement — APPROVED** [review]

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java
Comment (danthev) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java]: Done. Thanks!

**Review by johnjcasey — APPROVED** [review]
LGTM, though I think a lot of this is very low level encoding work, and should probably be pushed down to the Firestore library itself

**Pull request merged**
